File: docs/examples/user_types/ruby/02-user_classes.rb
# classes-start
require 'oso'

OSO ||= Oso.new

class Customer
  def initialize(id)
    @id = id
  end
end

OSO.register_class(Customer)

# internal-start
class InternalUser
  # ...

  def role
    db.query('SELECT role FROM internal_roles WHERE id = ?', id)
  end
end
# internal-end

OSO.register_class(InternalUser)

# account-start
class AccountManager < InternalUser
  def customer_accounts
    db.query('SELECT id FROM customer_accounts WHERE manager_id = ?', id)
  end
end
# account-end

# generate-start
def user_from_id(id)
  case db.query('SELECT type FROM users WHERE id = ?', id)
  when 'internal'
    actor = InternalUser.new(id: id)
    if actor.role == 'account_manager'
      AccountManager.new(id: id)
    else
      actor
    end
  when 'customer'
    Customer.new(id: id)
  end
end
# generate-end

File: docs/examples/user_types/nodejs/02-userClasses.js
// classes-start
const { Oso } = require('oso');

const oso = new Oso();

class Customer {
  constructor(id) {
    this.id = id;
  }
}

oso.registerClass(Customer);

// internal-start
class InternalUser {
  // ...

  role() {
    return db.query('SELECT role FROM internal_roles WHERE id = ?', this.id);
  }
}
// internal-end

oso.registerClass(InternalUser);

// account-start
class AccountManager extends InternalUser {
  customerAccounts() {
    return db.query(
      'SELECT id FROM customer_accounts WHERE manager_id = ?',
      this.id
    );
  }
}
// account-end

// generate-start
function userFromId(id) {
  const userType = db.query('SELECT type FROM users WHERE id = ?', id);
  if (userType === 'internal') {
    const actor = new InternalUser(id);
    if (actor.role() === 'account_manager') {
      return new AccountManager(id);
    } else {
      return actor;
    }
  } else if (userType === 'customer') {
    return new Customer(id);
  }
}
// generate-end

module.exports = { oso };
